src/reader.rs

```diff
@@ -138,52 +95,163 @@ impl<'a, R: Read> Reader<'a, R> {
-    fn read_block(&mut self) -> Result<(), Error> {
-        match decode(&Schema::Long, &mut self.reader) {
-            Ok(block) => {
-                if let Value::Long(block_len) = block {
-                    if let Value::Long(block_bytes) = decode(&Schema::Long, &mut self.reader)? {
-                        let mut bytes = vec![0u8; block_bytes as usize];
-                        self.reader.read_exact(&mut bytes)?;
-
-                        let mut marker = [0u8; 16];
-                        self.reader.read_exact(&mut marker)?;
-
-                        if marker != self.marker {
-                            return Err(DecodeError::new(
-                                "block marker does not match header marker",
-                            ).into())
-                        }
-
-                        self.codec.decompress(&mut bytes)?;
-
-                        self.items.clear();
-                        self.items.reserve_exact(block_len as usize);
-
-                        let mut block_bytes = &bytes[..];
-                        for _ in 0..block_len {
-                            let item = from_avro_datum(
-                                &self.writer_schema,
-                                &mut block_bytes,
-                                self.reader_schema,
-                            )?;
-                            self.items.push_back(item)
-                        }
-
-                        return Ok(())
-                    }
+    fn read_block_next(&mut self) -> Result<(), Error> {
+        if !self.is_empty() {
+            // I suppose this isn't technically an error...
+            panic!("(TODO Replace this with an error) Invalid state; shouldn't call read_block_next");
+        }
+        match util::read_long(&mut self.reader) {
+            Ok(block_len) => {
+                self.message_count = block_len as usize;
+                let block_bytes = util::read_long(&mut self.reader)?;
+                self.fill_buf(block_bytes as usize)?;
+                let mut marker = [0u8; 16];
+                self.reader.read_exact(&mut marker)?;
+
+                if marker != self.marker {
+                    return Err(DecodeError::new(
+                        "block marker does not match header marker",
+                    ).into())
                 }
+
+                // TODO (JAB): This is bad, but will fix later.
+                // Can use some "limited read" type to decode directly into the buffer.
+                self.codec.decompress(&mut self.buf)?;
+                
+                return Ok(())
             },
             Err(e) => match e.downcast::<::std::io::Error>()?.kind() {
                 // to not return any error in case we only finished to read cleanly from the stream
                 ErrorKind::UnexpectedEof => return Ok(()),
+                // TODO (JAB): This variant should probably return an error?
```

Comment:
TODO left here :)

Reply:
Yeah, that was a misunderstanding of what was happening in this code. It was fine how it is.
Removed the `TODO`